README.md
# CATALOG_GRID

This package provides a common template for Catalog Grids, using cards and other
Flutter widgets.

This project should be incremented as I personally feel the need to. If you have
a common usecase and feel like creating a template for it, please send me a PR.

[![Donate with Paypal](https://www.paypalobjects.com/en_US/i/btn/btn_donateCC_LG.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=B8ST4MMXWBQJ8&currency_code=BRL&source=url)

## Usage 
To use this package, add catalog_grid as a dependency in your pubspec.yaml file.

### Example
```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  List<Widget> images;
  List<Widget> contents;
  List<List<Widget>> options;

  @override
  Widget build(BuildContext context) {
    images ??= List.generate(
      8,
      (index) => Container(
        color: Colors.green,
        child: FlutterLogo(),
      ),
    );
    contents ??= List.generate(
      8,
      (index) => Text(
        'This is the example of a card content. '
        'Usually, a Text is used here. This is item $index content example.',
        textAlign: TextAlign.justify,
      ),
    );
    options = List.generate(
        8,
        (index) => index != 5
            ? [
                SizedBox(
                  width: 120,
                  child: RaisedButton.icon(
                    icon: Icon(Icons.delete),
                    label: Text("Remove"),
                    color: Colors.red,
                    textColor: Colors.white,
                    onPressed: () => print('Pressed delete'),
                  ),
                ),
                SizedBox(
                  width: 120,
                  child: RaisedButton.icon(
                    icon: Icon(Icons.edit),
                    label: Text("Edit"),
                    color: Colors.blue,
                    textColor: Colors.white,
                    onPressed: () => print('Pressed delete'),
                  ),
                ),
              ]
            : null);
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: Scaffold(
        appBar: AppBar(
          centerTitle: true,
          title: Text("Grid Catalog Example"),
        ),
        body: SafeArea(
          minimum: EdgeInsets.all(8.0),
          child: CatalogGrid(
            images: images,
            contents: contents,
            cardOptions: options,
            itemCount: 8,
            aspectRatio: 0.8,
            horizontalCount: 2,
          ),
        ),
      ),
    );
  }
}
```

## Status
This package is in it's initial stages. Helpful PRs are very welcome.


CHANGELOG.md
## [1.0.0] - Published to pub.dev


lib/catalog_card.dart
import 'package:flutter/material.dart';

class CatalogCard extends StatefulWidget {
  const CatalogCard({
    Key key,
    @required this.image,
    @required this.content,
    this.options,
  })  : assert(options == null || options is List<Widget>),
        super(key: key);

  final Widget image;
  final Widget content;
  final List<Widget> options;

  @override
  _CatalogCardState createState() => _CatalogCardState();
}

class _CatalogCardState extends State<CatalogCard> {
  @override
  Widget build(BuildContext context) {
    return Card(
      clipBehavior: Clip.antiAlias,
      shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(18.0)),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: <Widget>[
          Expanded(child: widget.image),
          Expanded(
            child: Padding(
              padding: const EdgeInsets.fromLTRB(16.0, 4.0, 16.0, 12.0),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.stretch,
                children: [
                  Expanded(child: SingleChildScrollView(child: widget.content)),
                  const SizedBox(height: 8.0),
                  widget.options != null
                      ? SizedBox(
                          height: 40,
                          child: ListView.separated(
                            shrinkWrap: true,
                            reverse: true,
                            scrollDirection: Axis.horizontal,
                            itemCount: widget.options.length,
                            itemBuilder: (context, i) => widget.options[i],
                            separatorBuilder: (context, j) =>
                                const SizedBox(width: 8.0),
                          ),
                        )
                      : Container(),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }
}


lib/catalog_grid.dart
library catalog_grid;

import './catalog_card.dart';
import 'package:flutter/material.dart';

/// A Grid Template for displaying items in cards.
class CatalogGrid extends StatefulWidget {
  CatalogGrid({
    Key key,
    @required this.images,
    @required this.contents,
    this.cardOptions,
    @required this.itemCount,
    this.childWidth = 300,
    this.horizontalCount,
    this.horizontalMargin = 8.0,
    this.verticalMargin = 8.0,
    this.aspectRatio = 0.6,
  })  : assert(itemCount != null && itemCount > 0),
        assert(childWidth != null || horizontalCount != null),
        assert(images.length == itemCount && contents.length == itemCount),
        assert(cardOptions == null ||
            cardOptions.isEmpty ||
            cardOptions.length == itemCount),
        super(key: key);

  /// List of Widgets to use in the top part of the card. If empty, use
  /// an empty [Container] to avoid mismatch of image and content.
  final List<Widget> images;

  /// List of Widgets to use in bottom part of the card. Usually a [Text]
  /// is used here. Use an empty Container or Text to avoid mismatch with
  /// images.
  final List<Widget> contents;

  /// List of actions to be added on the bottom of the card. This is shown
  /// below the content. Usually buttons are placed here.
  final List<List<Widget>> cardOptions;

  /// Total number of items in the Grid. Cannot be a negative number.
  final int itemCount;

  /// The width each child should have. Defaults to 300.
  final double childWidth;

  /// How many children should be in an horizontal "Line". If not null, [childWidth]
  /// would be ignored and width would be equal to screen's width / count.
  final int horizontalCount;

  /// Horizontal Spacing.
  final double horizontalMargin;

  /// Vertical Spacing.
  final double verticalMargin;

  /// Child Aspect Ratio. Defaults to 0.6.
  final double aspectRatio;

  @override
  _CatalogGridState createState() => _CatalogGridState();
}

class _CatalogGridState extends State<CatalogGrid> {
  @override
  Widget build(BuildContext context) {
    return GridView.builder(
      gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
        crossAxisCount: crossAxisCount(context),
        crossAxisSpacing: widget.horizontalMargin,
        mainAxisSpacing: widget.verticalMargin,
        childAspectRatio: widget.aspectRatio,
      ),
      itemCount: widget.itemCount,
      itemBuilder: (context, i) => CatalogCard(
        image: widget.images[i],
        content: widget.contents[i],
        options: widget.cardOptions[i],
      ),
    );
  }

  int crossAxisCount(BuildContext context) =>
      widget.horizontalCount ??
      (MediaQuery.of(context).size.width / widget.childWidth).floor();
}


example/lib/main.dart
import 'package:catalog_grid/catalog_grid.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  List<Widget> images;
  List<Widget> contents;
  List<List<Widget>> options;

  @override
  Widget build(BuildContext context) {
    images ??= List.generate(
      8,
      (index) => Container(
        color: Colors.green,
        child: FlutterLogo(),
      ),
    );
    contents ??= List.generate(
      8,
      (index) => Text(
        'This is the example of a card content. '
        'Usually, a Text is used here. This is item $index content example.',
        textAlign: TextAlign.justify,
      ),
    );
    options = List.generate(
        8,
        (index) => index != 5
            ? [
                SizedBox(
                  width: 120,
                  child: RaisedButton.icon(
                    icon: Icon(Icons.delete),
                    label: Text("Remove"),
                    color: Colors.red,
                    textColor: Colors.white,
                    onPressed: () => print('Pressed delete'),
                  ),
                ),
                SizedBox(
                  width: 120,
                  child: RaisedButton.icon(
                    icon: Icon(Icons.edit),
                    label: Text("Edit"),
                    color: Colors.blue,
                    textColor: Colors.white,
                    onPressed: () => print('Pressed delete'),
                  ),
                ),
              ]
            : null);
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: Scaffold(
        appBar: AppBar(
          centerTitle: true,
          title: Text("Grid Catalog Example"),
        ),
        body: SafeArea(
          minimum: EdgeInsets.all(8.0),
          child: CatalogGrid(
            images: images,
            contents: contents,
            cardOptions: options,
            itemCount: 8,
            aspectRatio: 0.8,
            horizontalCount: 2,
          ),
        ),
      ),
    );
  }
}
